src/main/java/com/metamx/cache/CaffeineCache.java

```diff
@@ -22,40 +22,56 @@
 import com.github.benmanes.caffeine.cache.Cache;
 import com.github.benmanes.caffeine.cache.Caffeine;
 import com.github.benmanes.caffeine.cache.stats.CacheStats;
+import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Maps;
+import com.google.common.primitives.Chars;
 import com.google.common.primitives.Ints;
 import com.metamx.common.logger.Logger;
 import com.metamx.emitter.service.ServiceEmitter;
 import com.metamx.emitter.service.ServiceMetricEvent;
 import net.jpountz.lz4.LZ4Compressor;
 import net.jpountz.lz4.LZ4Factory;
-import org.apache.commons.codec.digest.DigestUtils;
 
+import javax.annotation.Nullable;
 import java.nio.ByteBuffer;
 import java.util.Map;
+import java.util.concurrent.Executor;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 
 public class CaffeineCache implements io.druid.client.cache.Cache
 {
   private static final Logger log = new Logger(CaffeineCache.class);
+  private static final int FIXED_COST = 8; // Minimum cost in "weight" per entry;
   private final Cache<NamedKey, byte[]> cache;
   private final AtomicReference<CacheStats> priorStats = new AtomicReference<>(null);
 
 
   public static CaffeineCache create(final CaffeineCacheConfig config)
   {
+    return create(config, null);
+  }
+
+  // Used in testing
+  public static CaffeineCache create(final CaffeineCacheConfig config, @Nullable final Executor executor)
+  {
     Caffeine<Object, Object> builder = Caffeine.newBuilder().recordStats();
     if (config.getExpiration() >= 0) {
-      builder = builder
+      builder
           .expireAfterAccess(config.getExpiration(), TimeUnit.MILLISECONDS);
     }
     if (config.getMaxSize() >= 0) {
-      builder = builder
-          .maximumSize(config.getMaxSize())
-          .maximumWeight(config.getMaxSize());
+      builder
+          .maximumWeight(config.getMaxSize())
+          .weigher((NamedKey key, byte[] value) -> value.length
+                                                   + key.key.length
+                                                   + key.namespace.length() * Chars.BYTES
+                                                   + FIXED_COST);
+    }
+    if (executor != null) {
+      builder.executor(executor);
```

Comment:
Yes, or same-thread (`Runnable::run`) if you want it to amortize on the callers. The maintenance work is guarded by a state machine so only one execution is scheduled at any given time.

The lack of bounding recommendations is because the executor might be used for removal notifications, etc. If a `RejectedExecutionException` is thrown I do try to degrade gracefully by executing on the calling thread instead.

Since commonPool is intended to be shared (e.g. by Streams) it should be fairly safe to use. I use only O(1) algorithms so the execution time doesn't vary based on the size and stays short. I can understand the concern, but it might be unwarranted.